rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,466 @@
+extra to map these metric types to OTLP.
+
+Aggregated histograms are also natively supported, but they're simply called histograms in OTLP.
+
+#### Mapping sets
+
+Sets have no natural equivalent in OTLP. We would either need to drop support for them entirely, or
+follow the described path in the RFC of aggregating them at source level and emitting gauges
+directly.
+
+#### Mapping aggregated summaries
+
+Technically, the OpenTelemetry metrics data model
+[does support summaries](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#summary-legacy),
+but they are deprecated and not encouraged to be used, given that they lack the natural ability to
+be aggregated server-side.
+
+This is an area where we could either continue to support them natively in Vector -- with all of the
+logic we currently have to convert them to individual gauges where not supported by sinks -- or
+simply do the work to decompose them at the source-level and remove them as a metric type.
+
+#### Mapping distributions
+
+For the current `Distribution` metric type, we would map it to an OTLP `ExponentialHistogram`. This
+is the highest fidelity option for representing a wide range of sample values in a way that is
+mergeable.
+
+#### Mapping sketches
+
+Sketches have no natural equivalent in OTLP, at least as Vector supports them, which is only through
+a Datadog Agent-specific implementation of `DDSketch`. However, `DDSketch` itself is essentially a
+histogram under-the-hood, so we would be able to represent it as an OTLP histogram by simply
+exposing all sketch buckets directly.
+
+This would preserve the error guarantees of `DDSketch` itself as we would simply be representing it
+directly in terms of the buckets, but we would lose the useful properties such as being able to only
+track buckets that have values in them. While it might seem like a natural fit to map the `DDSketch`
+data structure to OTLP's exponential histogram, which functions the same in principle, a
+discontinuity between how both of them choose their bucket index mappings leaves Vector unable to
+losslessly convert one to the other.
+
+## Outstanding Questions
+
+- While we ostensibly look to fully support all systems that Vector provides components for, set
+  usage in StatsD is anecdotally almost nonexistent. How strong of an indicator would we want to see
+  before feeling comfortable removing support for sets entirely? This would imply not aggregating
+  them in the `statsd` source at all, as we would still remove `Set` from `MetricValue`.
+- Would we want to go as far as marking the removed/refactored metric types as deprecated, and then
+  remove support for them entirely from the Protocol Buffers schema in a future release? The
+  conversion logic (for going from the Protocol Buffers representation to the native Vector
+  representation) would represent additional technical debt that could hamper our efforts if we did
+  further refactoring in the future.
```

Comment:
Technically, if we follow our deprecation policy, we would be obligated to do this multi-step process.

Reply:
Yeah, it would absolutely need to happen in a multi-step fashion.

We would ideally batch this up as part of a 1.0 release change... where we deprecate the older variants now, pre-1.0, and then as part of a 1.0, we do the cleanup.